### DISCUSSION TITLE
tube filter with cell data as scalars

### DISCUSSION THREAD

**Discussion by ArashRabbani** [pyvista/pyvista#3730] [Q&A] [answered]
### Describe the bug, what's wrong, and what you expected.

Hi, 
I am trying to visualize a network of tubes with different sizes (a single tube radius should not change along its axis).
![image](https://user-images.githubusercontent.com/44697006/208254453-0c825e2c-627f-4fb2-882d-3f85a61b5986.png)

tube filter with cell data as scalars shows nothing. 
but tube filter with point data as scalars works fine. 

### Steps to reproduce the bug.

```python
import pyvista as pv
import numpy as np

polyline = pv.PolyData()
points = [[0.0, 0.0, 0.0], [10.0, 0.0, 0.0], [10.0, 10.0, 0.0], [0.0, 10.0, 0.0]]
polyline.points = points
the_cell=[2,0,1,2,1,2,2,2,3,2,3,0,2,0,2]
polyline.lines = the_cell

# it works fine with point data as scalar:
# polyline['value'] = np.arange(polyline.n_points)
# tube = polyline.tube(scalars='value', radius_factor=2,preference='point')
# tube.plot()

# but fails with cell data as scalar :
polyline['value'] = np.arange(polyline.n_cells)
tube = polyline.tube(scalars='value', radius_factor=2,preference='cell')
tube.plot()
```

### System Information

```shell
OS : Windows
            CPU(s) : 16
           Machine : AMD64
      Architecture : 64bit
               RAM : 15.6 GiB
       Environment : Python
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : NVIDIA GeForce RTX 3070/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 516.94

  Python 3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:12:32) [MSC
  v.1929 64 bit (AMD64)]

           pyvista : 0.37.0
               vtk : 9.2.2
             numpy : 1.23.5
           imageio : 2.22.4
            scooby : 0.7.0
             pooch : v1.6.0
        matplotlib : 3.6.2
         pyvistaqt : 0.9.0
             PyQt5 : 5.15.7
           IPython : 7.34.0
             scipy : 1.9.3
```

### Screenshots

scalars=cell data 
![image](https://user-images.githubusercontent.com/44697006/208254258-69e37087-2089-4345-a936-5f491126f91b.png)

<br>
<br>

scalars=point data 
![image](https://user-images.githubusercontent.com/44697006/208254284-b46168c0-7a7b-42c9-ba1c-105bee3ffe6c.png)

**Comment by banesullivan** [comment]
> a single tube radius should not change along its axis

If this is the case, then you do not want to use the `scalars` argument of the `tube()` filter but rather you want to specify the `scalars` when plotting.

**Comment by banesullivan** [comment, marked as answer]
```py
import pyvista as pv
import numpy as np

polyline = pv.PolyData()
points = [[0.0, 0.0, 0.0], [10.0, 0.0, 0.0], [10.0, 10.0, 0.0], [0.0, 10.0, 0.0]]
polyline.points = points
the_cell=[2,0,1,2,1,2,2,2,3,2,3,0,2,0,2]
polyline.lines = the_cell


# but fails with cell data as scalar :
polyline['value'] = np.arange(polyline.n_cells)
tube = polyline.tube()
tube.plot(scalars='value', preference='cell', notebook=False)
```
<img width="624" alt="Screen Shot 2022-12-19 at 11 54 19 AM" src="https://user-images.githubusercontent.com/22067021/208498971-cb2cf058-8a29-4f39-8d01-fed787fba2d8.png">

**Reply by ArashRabbani** [reply]
Thanks a lot for your reply. But this is not the intended functionality of having tubes with different sizes. For example, I am trying to plot 5 tubes with radii of 1 to 5 units. Is there any way that I can use the tube filter to do that? (The radius of every single tube should remain constant along its axis)

**Reply by banesullivan** [reply]
Ahh... I see what you're after now. We could do a better job of documenting this, but the `scalars` argument to the `tube()` filter only works for POINT data... AFAIK. Ref https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.tube.html#tube

What you want to do is still achievable though. Here is one way to do it by iterating over each cell (each line that will become a tube) and specifying the `radius` when calling the `tube()` filter

```py
import pyvista as pv
import numpy as np

polyline = pv.PolyData()
points = [[0.0, 0.0, 0.0], [10.0, 0.0, 0.0], [10.0, 10.0, 0.0], [0.0, 10.0, 0.0]]
polyline.points = points
polyline.lines = [2,0,1,2,1,2,2,2,3,2,3,0,2,0,2]
# Add radius to each cell of lines
polyline['value'] = np.arange(polyline.n_cells, dtype=float) + 1  # +1 to avoid 0-radius
polyline['value'] *=  0.1  # scaling
# Split mesh by each cell and use cell data to control radius
tubes = pv.PolyData()
for cid in range(polyline.n_cells):
    cell = polyline.extract_cells([cid]).extract_geometry()
    tubes += cell.tube(radius=cell['value'])
tubes.plot(notebook=False)
```
<img width="624" alt="Screen Shot 2022-12-19 at 7 20 18 PM" src="https://user-images.githubusercontent.com/22067021/208565819-3b12fc05-8155-4ff1-a8cc-33f999a94b97.png">

**Reply by ArashRabbani** [reply]
Thanks a lot for your reply. That is exactly it. I was wondering if you recommend this approach to be used for plotting around 10K to 20K tubes, because of the loop.

**Reply by banesullivan** [reply]
> I was wondering if you recommend this approach to be used for plotting around 10K to 20K tubes, because of the loop.

Hm that's a good point... You may run into performance issues. 

Another solution may be to have one mesh with duplicate points. Then have POINT data of the radius on each point for a given cell

**Reply by ArashRabbani** [reply]
That is really a smart trick, thanks for sharing. 
BTW, your work is great on Pyvista, big fan.

**Reply by banesullivan** [reply]
I thought this might be fun to solve, so here is an example with duplicated points and no for-loop:

```py
import pyvista as pv
import numpy as np

points = np.array([
    [ 0.,  0.,  0.], # 0
    [ 0.,  0.,  0.], # 1
    [ 0.,  0.,  0.], # 2
    [10.,  0.,  0.], # 3
    [10.,  0.,  0.], # 4
    [10., 10.,  0.], # 5
    [10., 10.,  0.], # 6
    [10., 10.,  0.], # 7
    [ 0., 10.,  0.], # 8
    [ 0., 10.,  0.], # 9
])

polyline = pv.PolyData()
polyline.points = points
polyline.lines = [2,0,3, #
                  2,4,6, #
                  2,7,8, #
                  2,9,1, #
                  2,2,5, #
                 ]



# Add radius to each cell of lines
polyline['value'] = np.arange(polyline.n_cells, dtype=float) + 1  # +1 to avoid 0-radius
polyline['value'] *=  0.1  # scaling
# ProTip: cell_to_point data filter to migrate radii to points
polyline = polyline.ctp()
tubes = polyline.tube(scalars='value', absolute=True)

tubes.plot(notebook=0)
```
![Screen Shot 2022-12-20 at 12 05 46 PM](https://user-images.githubusercontent.com/22067021/208746817-3ff6babc-7cb4-4743-8295-0013caa96c3c.png)

**Reply by ArashRabbani** [reply]
Amazing, and the performance was very good on a high number of tubes. Thanks
